src/libraries/System.Private.CoreLib/src/System/Environment.OSX.cs

```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
+using nint = System.IntPtr;
+
+namespace System
+{
+    public static partial class Environment
+    {
+        private static OperatingSystem GetOSVersion()
+        {
+            int major = 0;
+            int minor = 0;
+            int patch = 0;
+
+            IntPtr processInfo = objc_msgSend(objc_getClass("NSProcessInfo"), sel_getUid("processInfo"));
+            if (processInfo != IntPtr.Zero)
+            {
+                NSOperatingSystemVersion osVersion = get_operatingSystemVersion(processInfo, sel_getUid("operatingSystemVersion"));
+                major = osVersion.majorVersion.ToInt32();
+                minor = osVersion.minorVersion.ToInt32();
+                patch = osVersion.patchVersion.ToInt32();
+            }
+
+            // For compatibility reasons with Mono, PlatformID.Unix is returned on MacOSX. PlatformID.MacOSX
+            // is hidden from the editor and shouldn't be used.
+            return new OperatingSystem(PlatformID.Unix, new Version(major, minor, patch));
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct NSOperatingSystemVersion
+        {
+            public nint majorVersion;
+            public nint minorVersion;
+            public nint patchVersion;
+        }
+
+        [DllImport("libobjc.dylib")]
+        private static extern IntPtr objc_getClass(string className);
+        [DllImport("libobjc.dylib")]
+        private static extern IntPtr sel_getUid(string selector);
+        [DllImport("libobjc.dylib")]
+        private static extern IntPtr objc_msgSend(IntPtr basePtr, IntPtr selector);
+        [DllImport("libobjc.dylib", EntryPoint = "objc_msgSend_stret")]
+        private static extern NSOperatingSystemVersion get_operatingSystemVersion(IntPtr basePtr, IntPtr selector);
```

Comment:
Why does this one have a different name than the entry point?

Reply:
The way `objc_msgSend_stret` works (from my understanding of reading docs and examples) is that the arguments and return values are mutable depending on the function's signature. The way I've seen C code calling it is by defining a function pointer with the right signature and casting `objc_msgSend_stret` to it.

I named the function `get_operatingSystemVersion` since the return value is specific to OS Version.

See https://blog.lazerwalker.com/objective-c,/code/2013/10/12/the-objective-c-runtime-and-objc-msgsend-stret.html

> The solution? You need to cast the objc_msgSend_stret function itself into a function whose return type is the struct you’re expecting. Here is a complete, functioning version of our example:

```
UIView *view = [[UIView alloc] initWithFrame:CGRectZero];

CGRect (*sendRectFn)(id receiver, SEL operation);
sendRectFn = (CGRect(*)(id, SEL))objc_msgSend_stret;
CGRect frame = sendRectFn(view, @selector(frame));
```